File: spark-scala/src/main/java/org/guoqing/sparkscala/fast_data/KeyValueOption.scala
package org.guoqing.sparkscala

import org.guoqing.sparkscala.config.InitSpark

/**
  * $DESCRIPTION
  *
  * @author DGQ
  * @since 2019/3/4
  *
  */

case class Store(string: String)

object KeyValueOption {

  val sc = InitSpark.getSparkConf()

  def main(args: Array[String]): Unit = {
    /* 键值对的简单操作
    val lines = sc.parallelize(List("you are ", "nice"))
    val pairs = lines.map(x => (x.split(" ")(0), x))
    pairs.map { case (x, y) => x }.foreach(println(_))
    */
    //值对映射
    /*val pairs = sc
      .parallelize(List(("panda", 0),
        ("pink", 3), ("pirate", 3),
        ("panda", 1), ("pink", 4)))
    pairs.mapValues(x => (x, 1)).reduceByKey((x, y) => (x._1 + y._1, x._2 + y._2))
      .foreach(println(_))*/
    //wordCounts()
    //average()
    testJoin()
  }

  /**
    * 单词计数
    */
  def wordCounts(): Unit = {
    val lines = sc.parallelize(List
    ("I love ning huizhong.",
      "I don't know when she walks into my life, into my heart . when do i walks into heart ?"))

    val words = lines.flatMap(t => t.split(" "))
    val result = words.map(t => (t, 1)).reduceByKey((x, y) => x + y)
    result.foreach(println(_))
  }

  /**
    * 使用combineByKey()求每个键对应的平均值
    */
  def average(): Unit = {
    val input = List(("A", 5), ("B", 3), ("G", 3), ("A", 4), ("A", 7), ("B", 9))
    val rdd = sc.parallelize(input)

    val result = rdd.combineByKey(v => (v, 1),
      (acc: (Int, Int), v) => (acc._1 + v, acc._2 + 1),
      (acc1: (Int, Int), acc2: (Int, Int)) => (acc1._1 + acc2._1, acc1._2 + acc2._2)).
      map { case (key, value) => (key, value._1 / value._2.toFloat) }
    result.collectAsMap().map(println(_))
  }

  def testJoin(): Unit = {
    val storeAddress = sc.parallelize(Seq((Store("Ritual"), "1026 Valencia St"),
      (Store("  "), "748 Van Ness Ave"),
      (Store("Philz"), "3101 24th St"),
      (Store("Starbucks"), "Seattle")))
    val storeRating = sc.parallelize(Seq((Store("Ritual"), 4.9), (Store("Philz"), 4.8)))
    val result = storeAddress.join(storeRating)
    result.foreach(t => println(t._1 + ":" + t._2))

  }
}

File: spark-java/src/main/java/org/guoqing/sparkjava/fast/KeyValueOption.java
package org.guoqing.sparkjava.fast;

import org.apache.commons.lang3.StringUtils;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.guoqing.sparkjava.config.InitSpark;
import scala.Serializable;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * 键值对操作
 *
 * @author dengguoqing
 * @date 2019/3/4
 */
public class KeyValueOption {
    private static JavaSparkContext sc = InitSpark.sparkContext();
    private static JavaPairRDD<String, String> pairRDD = sc
            .parallelize(Arrays.asList("I love the girl", "that is ning"))
            .mapToPair(x -> new Tuple2<>(StringUtils.split(x, " ")[0], x));

    public static void main(String[] args) {
        //createPairRDD();
        //filterSecondValue();
        //countWords();

        average();
    }

    /**
     * 创建Pair RDD
     */
    public static void createPairRDD() {
        pairRDD.map(t -> t._1 + t._2).foreach(x -> System.out.println(x));
    }

    /**
     * 根据第二个值过滤数据
     */
    public static void filterSecondValue() {
        JavaPairRDD<String, String> ning = pairRDD
                .filter(t -> t._2.contains("ning"));
        ning.map(t -> t._2).foreach(t -> System.out.println(t));
    }

    public static void countWords() {
        JavaRDD<String> input = sc.parallelize(
                Arrays.asList("I love ning huizhong.",
                        "I don't know when she walks into my life, into my heart . when do i walks into heart ?"));
        JavaRDD<String> words = input
                .flatMap(x -> Arrays.asList(x.split(" ")).iterator())
                .filter(x -> x.matches("\\w+")).map(String::toUpperCase);
        JavaPairRDD<String, Integer> pairRDD = words
                .mapToPair(x -> new Tuple2<>(x, 1))
                .reduceByKey((x, y) -> x + y);
        pairRDD.foreach(t -> System.out.println(t._1 + " " + t._2));

    }

    private static Function<Integer, AvgCount> createAcc = x -> new AvgCount(x,
            1);

    private static Function2<AvgCount, Integer, AvgCount> addAndCount = (a, x) -> new AvgCount(
            a.total + x, a.num + 1);

    private static Function2<AvgCount, AvgCount, AvgCount> combine = (a, b) -> new AvgCount(
            a.total + b.total, a.num + b.num);


    /**
     * 使用combine()求每个键对应的平均值
     */
    public static void average() {
        List<Tuple2<String, Integer>> input = Arrays
                .asList(new Tuple2<>("A", 5), new Tuple2<>("B", 3),
                        new Tuple2<>("G", 3), new Tuple2<>("A", 4),
                        new Tuple2<>("A", 7), new Tuple2<>("B", 9));
        JavaRDD<Tuple2<String, Integer>> rdd = sc.parallelize(input);
        JavaPairRDD<String, Integer> nums = rdd.mapToPair(x -> x);
        JavaPairRDD<String, AvgCount> result = nums
                .combineByKey(createAcc, addAndCount, combine);

        Map<String, AvgCount> countMap = result.collectAsMap();
        countMap.forEach(
                (key, value) -> System.out.println(key + ":" + value.avg()));
    }


    public static class AvgCount implements Serializable {
        public int total;
        public int num;

        public AvgCount(int total, int num) {
            this.total = total;
            this.num = num;
        }

        public float avg() {
            return total / (float) num;
        }

    }
}
